src/consensus/params.cpp

```diff
@@ -36,32 +40,138 @@ namespace Consensus {
+        //
         // Halving(h) is defined as floor(f(h)) where f is a strictly increasing rational
         // function, so it's sufficient to solve for f(height) = 1 in the rationals and
         // then take ceiling(height - 1).
-        // H := blossom activation height; SS := SubsidySlowStartShift(); R := BLOSSOM_POW_TARGET_SPACING_RATIO
+        //
+        // H := blossom activation height; 
+        // SS := SubsidySlowStartShift(); 
+        // R := 1 / (postInterval / preInterval) = BLOSSOM_POW_TARGET_SPACING_RATIO
+        // (The following calculation depends on BLOSSOM_POW_TARGET_SPACING_RATIO being an integer.) 
+        //
         // preBlossom:
         // 1 = (height - SS) / preInterval
         // height = preInterval + SS
+        //
         // postBlossom:
         // 1 = (H - SS) / preInterval + (height - H) / postInterval
         // height = H + postInterval - (H - SS) * (postInterval / preInterval)
         // height = H + postInterval - (H - SS) * R
+        //
         // Note: This depends on R being an integer
-        bool blossomActive = NetworkUpgradeActive(nHeight, Consensus::UPGRADE_BLOSSOM);
-        if (blossomActive) {
+        if (NetworkUpgradeActive(nHeight, Consensus::UPGRADE_BLOSSOM)) {
             int blossomActivationHeight = vUpgrades[Consensus::UPGRADE_BLOSSOM].nActivationHeight;
-            // The following calculation depends on BLOSSOM_POW_TARGET_SPACING_RATIO being an integer. 
-            return blossomActivationHeight + nPostBlossomSubsidyHalvingInterval
-                - (blossomActivationHeight - SubsidySlowStartShift()) * BLOSSOM_POW_TARGET_SPACING_RATIO - 1;
+            return blossomActivationHeight 
+                - ((blossomActivationHeight - SubsidySlowStartShift()) * BLOSSOM_POW_TARGET_SPACING_RATIO)
+                + (nPostBlossomSubsidyHalvingInterval * (halvingIndex - Halving(blossomActivationHeight)));
         } else {
-            return nPreBlossomSubsidyHalvingInterval + SubsidySlowStartShift() - 1;
+            return (halvingIndex * nPreBlossomSubsidyHalvingInterval) + SubsidySlowStartShift();
         }
     }
 
+    int Params::GetLastFoundersRewardBlockHeight(int nHeight) const {
+        return HalvingHeight(nHeight, 1) - 1;
+    }
+
+    int Params::FundingPeriodIndex(int fundingStreamStartHeight, int nHeight) const {
+        int firstHalvingHeight = HalvingHeight(fundingStreamStartHeight, 1);
+
+        // If the start height of the funding period is not aligned to a multiple of the 
+        // funding period length, the first funding period will be shorter than the
+        // funding period length. 
+        auto startPeriodOffset = (fundingStreamStartHeight - firstHalvingHeight) % nFundingPeriodLength;
+        if (startPeriodOffset < 0) startPeriodOffset += nFundingPeriodLength; // C++ '%' is remainder, not modulus!
+
+        return (nHeight - fundingStreamStartHeight + startPeriodOffset) / nFundingPeriodLength;
```

Comment:
I'm almost satisfied that this is coincident with the behaviour specified in ZIP 207. I'll stare at it some more.

Reply:
I'm now convinced that this has the correct behaviour:
```
            startPeriodOffset
                 |-----|
...-----||-------|-----==|=======|=======|===--...
     Halving1          |--------------------|
            fundingStreamStartHeight     nHeight
```